### PR TITLE
Improvements to the compiler

### PR DESCRIPTION
**Pull request by michalmuskala**
This is several loosely related changes to the compiler that either clean things up or improve performance. With all the changes, on some benchmarks the compiler performance improved about 5%.

### CODE REVIEW

File: lib/elixir/src/elixir_quote.erl [hunk 1]
```diff
@@ -418,7 +430,8 @@ keystore(_Key, Meta, nil) ->
 keystore(Key, Meta, Value) ->
   lists:keystore(Key, 1, Meta, {Key, Value}).
 keynew(Key, Meta, Value) ->
-  case keyfind(Key, Meta) of
-    {Key, _} -> Meta;
-    _ -> keystore(Key, Meta, Value)
+  case lists:keymember(Key, 1, Meta) of
+    true -> Meta;
+    false when Value =:= nil -> Meta;
```
Comment (josevalim) [inline lib/elixir/src/elixir_quote.erl]: It seems we always return `Meta` when the `Value` is `nil`. When is the value `nil` in the first place?

Comment (josevalim) [reply lib/elixir/src/elixir_quote.erl]: I can't identify any case where the value would be `nil`.

Comment (michalmuskala) [reply lib/elixir/src/elixir_quote.erl]: I just inlined the code from keystore that explicitly checks for the value being `nil`. Maybe it has some other uses there, but here it is useless?

Comment (josevalim) [reply lib/elixir/src/elixir_quote.erl]: @michalmuskala so maybe we should move the `nil` check up, like we do for `keystore`. Or remove it altogether. Your call.

Comment (michalmuskala) [reply lib/elixir/src/elixir_quote.erl]: Yeah, it seems `keynew` is only ever called with either a `line` or `counter` and a value of `nil` makes no sense in there. I'll remove the check.

**Pull request merged**